### PR TITLE
git.detached: Fix traceback when rev is a SHA and is not present locally

### LINKED ISSUE #43143: git.detached does not fetch if rev is missing from local
### Description of Issue/Question
When applying the `git.detached` state with a `rev` specified, if the `rev` is set to a commit hash which is available on the remote but not locally, the state fails with an exception.

### Setup
Example state file:
```
# git-state.sls
https://github.com/saltstack/salt.git:
  git.detached:
    - rev: '5cb40aca4269174ff633f66ec76fb56b6880d2ae'
    - target: '/var/lib/salt/git-clones/salt'
```

Repo has already been cloned to the target but a new commit has been added to the remote since the clone, the state is specifying that new commit hash in `rev`.

### Steps to Reproduce Issue
With the setup above, apply state:
```
# salt-call -l debug state.apply master.external-modules
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/beacons.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/beacons.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/engine.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/engine.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/reactor.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/reactor.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: <host>.<domain>
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/beacons.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/beacons.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/engine.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/engine.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/reactor.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/reactor.conf
[DEBUG   ] Please install 'virt-what' to improve results of the 'virtual' grain.
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (1030)
[DEBUG   ] Setting zmq_reconnect_ivl to '1030ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] SaltEvent PUB socket URI: /var/run/salt/minion/minion_event_d9754c0b30_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: /var/run/salt/minion/minion_event_d9754c0b30_pull.ipc
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/minion/minion_event_d9754c0b30_pull.ipc
[DEBUG   ] Sending event: tag = salt/auth/creds; data = {'_stamp': '2017-08-23T13:32:48.753947', 'creds': {'publish_port': 4505, 'aes': '<redacted>', 'master_uri': 'tcp://127.0.0.1:4506'}, 'key': ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')}
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Determining pillar cache
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded state.apply
[DEBUG   ] LazyLoaded saltutil.is_running
[DEBUG   ] LazyLoaded grains.get
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] In saltenv 'base', looking at rel_path 'git-state.sls' to resolve 'salt://git-state.sls'
[DEBUG   ] In saltenv 'base', ** considering ** path '/var/cache/salt/minion/files/base/git-state.sls' to resolve 'salt://git-state.sls'
[DEBUG   ] compile template: /var/cache/salt/minion/files/base/git-state.sls
[DEBUG   ] Jinja search path: ['/var/cache/salt/minion/files/base']
[PROFILE ] Time (in seconds) to render '/var/cache/salt/minion/files/base/git-state.sls' using 'jinja' renderer: 0.00683093070984
[DEBUG   ] Rendered data from file: /var/cache/salt/minion/files/base/git-state.sls:
# git-state.sls
https://github.com/saltstack/salt.git:
  git.detached:
    - rev: '5cb40aca4269174ff633f66ec76fb56b6880d2ae'
    - target: '/var/lib/salt/git-clones/salt'
[DEBUG   ] LazyLoaded config.get
[DEBUG   ] Results of YAML rendering: 
OrderedDict([('https://github.com/saltstack/salt.git', OrderedDict([('git.detached', [OrderedDict([('rev', '5cb40aca4269174ff633f66ec76fb56b6880d2ae')]), OrderedDict([('target', '/var/lib/salt/git-clones/salt')])])]))])
[PROFILE ] Time (in seconds) to render '/var/cache/salt/minion/files/base/git-state.sls' using 'yaml' renderer: 0.0134608745575
[DEBUG   ] LazyLoaded git.version
[DEBUG   ] LazyLoaded cmd.run_all
[INFO    ] Executing command ['git', '--version'] in directory '/root'
[DEBUG   ] stdout: git version 1.8.3.1
[DEBUG   ] LazyLoaded git.detached
[INFO    ] Running state [https://github.com/saltstack/salt.git] at time 13:32:49.637751
[INFO    ] Executing state git.detached for [https://github.com/saltstack/salt.git]
[INFO    ] Checking local revision for /var/lib/salt/git-clones/salt
[INFO    ] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/var/lib/salt/git-clones/salt'
[DEBUG   ] stdout: 1280c03c31599ed89d755277478bdcf79b03befb
[INFO    ] Checking local branch for /var/lib/salt/git-clones/salt
[INFO    ] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/var/lib/salt/git-clones/salt'
[DEBUG   ] stdout: HEAD
[INFO    ] Executing command ['git', 'describe', '--always', '5cb40aca4269174ff633f66ec76fb56b6880d2ae'] in directory '/var/lib/salt/git-clones/salt'
[ERROR   ] Command '['git', 'describe', '--always', '5cb40aca4269174ff633f66ec76fb56b6880d2ae']' failed with return code: 128
[ERROR   ] stderr: fatal: 5cb40aca4269174ff633f66ec76fb56b6880d2ae is not a valid 'commit' object
[ERROR   ] retcode: 128
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/state.py", line 1837, in call
    **cdata['kwargs'])
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1794, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/states/git.py", line 2236, in detached
    password=password):
  File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 1557, in describe
    ignore_retcode=ignore_retcode)['stdout']
  File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 320, in _git_run
    raise CommandExecutionError(msg)
CommandExecutionError: Command 'git describe --always 5cb40aca4269174ff633f66ec76fb56b6880d2ae' failed: fatal: 5cb40aca4269174ff633f66ec76fb56b6880d2ae is not a valid 'commit' object

[INFO    ] Completed state [https://github.com/saltstack/salt.git] at time 13:32:49.744547 duration_in_ms=106.805
[DEBUG   ] File /var/cache/salt/minion/accumulator/62923920 does not exist, no need to cleanup.
[DEBUG   ] LazyLoaded config.option
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[DEBUG   ] LazyLoaded highstate.output
local:
----------
          ID: https://github.com/saltstack/salt.git
    Function: git.detached
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/site-packages/salt/state.py", line 1837, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1794, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/states/git.py", line 2236, in detached
                  password=password):
                File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 1557, in describe
                  ignore_retcode=ignore_retcode)['stdout']
                File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 320, in _git_run
                  raise CommandExecutionError(msg)
              CommandExecutionError: Command 'git describe --always 5cb40aca4269174ff633f66ec76fb56b6880d2ae' failed: fatal: 5cb40aca4269174ff633f66ec76fb56b6880d2ae is not a valid 'commit' object
     Started: 13:32:49.637742
    Duration: 106.805 ms
     Changes:   

Summary for local
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time: 106.805 ms
```

### Versions Report
```
$ salt --versions-report
Salt Version:
           Salt: 2017.7.1
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: 1.10.6
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: 0.24.6
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: 0.24.2
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.6
 
System Versions:
           dist: centos 7.3.1611 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-514.16.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core
```

### PR DESCRIPTION
**Pull request by terminalmage**
This catches the (expected) CommandExecutionError when git.describe is
run on a commit that doesn't exist.

This also renames the remote_ref_type variable to remote_rev_type. This
change was made for the 2017.7 branch, but this fix is being applied to
2016.11, so by making the rename here as well we will avoid a potential
bug from creeping in on a merge-forward.

Resolves #43143.

### CODE REVIEW

File: salt/states/git.py [hunk 1]
```diff
@@ -2238,13 +2238,18 @@ def detached(name,
...
-                                             password=password):
-            # The ref is a hash and it exists locally so skip to checkout
-            hash_exists_locally = True
+        if remote_rev_type is 'hash':
+            try:
+                __salt__['git.describe'](target,
+                                         rev,
+                                         user=user,
+                                         password=password,
+                                         ignore_retcode=True):
```
Comment (abulford) [inline salt/states/git.py]: I think you forgot to remove the colon here.

File: salt/states/git.py [hunk 2]
```diff
@@ -2238,13 +2238,18 @@ def detached(name,
...
-                and __salt__['git.describe'](target,
-                                             ref,
-                                             user=user,
-                                             password=password):
-            # The ref is a hash and it exists locally so skip to checkout
-            hash_exists_locally = True
+        if remote_rev_type is 'hash':
+            try:
+                __salt__['git.describe'](target,
+                                         rev,
```
Comment (abulford) [inline salt/states/git.py]: In this branch `ref` is still used, so this variable isn't declared - did you mean to put this in a different branch?

**Comment by terminalmage** [comment]
@abulford thanks for the review, you are correct on both counts. The `ref` variable was renamed too for 2017.7.

@rallytime this will need to be kept in mind for the merge-forward. The variable name in the `git.describe` will need to be `rev`.

**Review by rallytime — APPROVED** [review]

**Comment by rallytime** [comment]
Thanks for the heads-up. That helped with the merge conflict resolution in #43201.

**Pull request merged**